### PR TITLE
feat(api, step-generation): emit and ingest command keys

### LINKED ISSUE #10799: bug: Run log does not track steps of a JSONv6 protocol
## Overview

When running a Python Protocol API v3 protocol or a JSON v6 protocol, the Opentrons App's run log does not update itself to show which command is currently running.

## Steps to reproduce

Run this JSONv6 protocol:  [JSON v6 Test Protocol.json.zip](https://github.com/Opentrons/opentrons/files/8961467/JSON.v6.Test.Protocol.json.zip)

Tested on 6.0.0-beta.0 and `edge` (as of 2022-06-22)

## Current behavior

The app shows steps from the protocol's analysis, but does not show the steps actually executed in the run.

Screen Shot 2022-06-16 at 12 17 11 PM

Pausing, playing, or letting the run finish don't appear to have any effect.

## Expected behavior

As with Python Protocol API v2 protocols:

* The current step should be highlighted in blue.
* The start and finish time of each step should be shown.

## Implemenatation notes

- Protocol Designer should attach unique `key` fields to each command
    - `key` is already present in JSON Protocol Schema v6
- The ProtocolEngine should accept an optional `key` parameter as part of the `CommandCreate` request object
- If the `CommandCreate` object has a `key`, it should be used and returned by the HTTP API

## Additional notes

This is also true of PAPIv3 protocols, but more-or-less expected because we don't support command keys in native ProtocolEngine commands yet. However, behavior of the run log UI in a PAPIv3 context is still problematic, because the run log does not update even as the `Run` resource tells the app that commands are executing.

[Dumps of HTTP responses during a PAPIv3 run](https://gist.github.com/SyntaxColoring/9df466c9bf0135ceabb02496e6b21f57). Note that some commands are missing because this was taken midway through the protocol, when the protocol was paused.

### PR DESCRIPTION
**Pull request by shlokamin**
# Overview
Emit keys in step generation and consume the in the api. This fixes the issue where the run log does not update itself to show which command is currently running.

closes #10799.

# Changelog
- Removed command_key from QueueCommandAction
- Added key to BaseCommandCreate
- ProtocolEngine:handle_action() changed assignment for key to use key if exists
- json_command_translator added key assignment
- Updated involving tests
- Emit command keys via uuid in all atomic command creators

# Review requests

1. are we missing any tests? 
2. added docstring's make sense?
3. tested on a robot with PD. works as expected!
4. uploading python protocols
5. create a v6 protocol from [edge PD](http://sandbox.designer.opentrons.com/edge/), upload it to the robot and make sure the run log works as expected

# Risk assessment

Medium. effects uploading Json v6 protocols and their analysis. added optional key to all commands, make sure nothing else is effected.

### CODE REVIEW

File: api/src/opentrons/protocol_engine/protocol_engine.py [hunk 1]
```diff
@@ -136,9 +136,6 @@ def add_command(self, request: CommandCreate) -> Command:
             QueueCommandAction(
                 request=request,
                 command_id=command_id,
-                # TODO(mc, 2021-12-13): generate a command key from params and state
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_engine/protocol_engine.py:146]: @mcous is it ok I removed this? I figured since we are now consuming the front end key we might not need this anymore

File: api/src/opentrons/protocol_engine/commands/command.py [hunk 2]
```diff
@@ -76,6 +76,12 @@ class BaseCommandCreate(GenericModel, Generic[CommandParamsT]):
             " and module setup, like pre-heating."
         ),
     )
+    key: Optional[str] = Field(
+        None,
+        description=(
+            "UI command key. This field will help" " protocol analysis for JSON v6."
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_engine/commands/command.py]:
```suggestion
            "Command key generated by UI and consumed by the api. By adding this field we are enabeling " " protocol analysis for JSON v6."
```

File: robot-server/tests/integration/http_api/runs/test_pause_then_cancel.tavern.yaml [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    request:
+      url: '{host:s}:{port:d}/runs/{run_id}/actions'
+      method: POST
+      json:
+        data:
+          actionType: pause
+    response:
+      status_code: 201
+
+#  - name: Wait for the delay to complete
```
Comment (TamarZanzouri) [inline robot-server/tests/integration/http_api/runs/test_pause_then_cancel.tavern.yaml]: @mcous left this here just to confirm we don't need it

File: api/src/opentrons/protocol_engine/commands/command.py [hunk 4]
```diff
@@ -76,6 +76,13 @@ class BaseCommandCreate(GenericModel, Generic[CommandParamsT]):
             " and module setup, like pre-heating."
         ),
     )
+    key: Optional[str] = Field(
+        None,
+        description=(
+            "Command key generated by UI and consumed by the api. By adding this field we are enabeling "
+            " protocol analysis for JSON v6."
+        ),
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/commands/command.py]: Remember that the context of these description is that of a HTTP API user, not of an Opentrons developer.

```suggestion
        description=(
            "A key value, unique in this run, that can be used to track"
            " the same logical command across multiple runs of the same protocol."
            " If a value is not provided, one will be generated."
        ),
```

File: api/src/opentrons/protocol_engine/state/commands.py [hunk 5]
```diff
@@ -181,7 +181,9 @@ def handle_action(self, action: Action) -> None:  # noqa: C901
             # For now, unit tests cover mapping every request type
             queued_command = action.request._CommandCls.construct(
                 id=action.command_id,
-                key=action.command_key,
+                key=action.request.key
+                if action.request.key is not None
+                else action.command_id,
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/state/commands.py]: Can we use parentheses for readability and preserve the TODO from `protocol_engine.py`?

```suggestion
                key=(
                    action.request.key
                    if action.request.key is not None
                    # TODO(mc, 2021-12-13): generate a command key from params and state
                    # https://github.com/Opentrons/opentrons/issues/8986
                    else action.command_id
                ),
```

File: api/src/opentrons/protocol_engine/state/commands.py [hunk 6]
```diff
@@ -500,7 +502,7 @@ def get_all_complete(self) -> bool:
             CommandExecutionFailedError: if any added command failed.
         """
         no_command_running = self._state.running_command_id is None
-        no_command_queued = len(self._state.queued_command_ids) == 0
+        no_command_queued = self.get_next_queued() is None
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/state/commands.py]: I don't see a unit test for this change. Are we missing the test or am I misreading?

File: robot-server/tests/integration/http_api/runs/test_pause_then_cancel.tavern.yaml (hunk 3 above)
Comment (mcous) [reply robot-server/tests/integration/http_api/runs/test_pause_then_cancel.tavern.yaml]: We definitely need it, please revert. If you comment this step out and run the test on `edge`, it will pass. This step ensures the test will fail on edge and catches the bug we observed.

The bug was that a run like this cannot be stopped:

1. Command 1: `status: succeeded`
2. Command 2: `status: queued`

Removing this block changes the test to do this instead:

1. Command 1: `status: running`
2. Command 2: `status: queued`

This second scenario was not exhibiting buggy behavior. If anything we probably need both scenarios in this test file. Could you make that addition?

File: step-generation/src/utils/index.ts [hunk 7]
```diff
@@ -1,17 +1,21 @@
+import uuidv1 from 'uuid/v1'
```
Comment (mcous) [inline step-generation/src/utils/index.ts]: I think we should be using `uuid/v4`. It's more common than v1 in our codebase, and ever so slightly more "safe" (only in the most pedantic technical sense) choice given our use of web workers across the front-end stack

(v1 generates IDs using the system clock and system MAC address, while v4 is "random". You could technically collide v1 IDs with multiple ID producers on the same machine, though the real-life risk is minimal to non-existent in our current use of UUIDs)

**Review by mcous — CHANGES_REQUESTED** [review]
I think we're missing a unit test and the change to the tavern acceptance test causes it to miss the cancellation bug

File: api/src/opentrons/protocol_engine/state/commands.py [hunk 8]
```diff
@@ -296,6 +302,7 @@ def handle_action(self, action: Action) -> None:  # noqa: C901
             if not self._state.run_result:
                 self._state.queue_status = QueueStatus.PAUSED
                 self._state.run_result = RunResult.STOPPED
+                self._state.queued_command_ids.clear()
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/state/commands.py:312]: What does this do?

I see the unit test for it, but I don't understand the higher-level intent.

File: robot-server/tests/integration/http_api/runs/test_pause_then_cancel.tavern.yaml [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+      url: '{host:s}:{port:d}/runs/{run_id}/commands'
+      method: GET
+    response:
+      status_code: 200
+      strict:
+        - json:off
+      json:
+        data:
+          - status: succeeded
+          - status: queued
```
Comment (SyntaxColoring) [inline robot-server/tests/integration/http_api/runs/test_pause_then_cancel.tavern.yaml:67]: What do two `status`es mean here?

File: robot-server/tests/integration/http_api/runs/test_pause_then_cancel.tavern.yaml [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    request:
+      url: '{host:s}:{port:d}/runs/{run_id}/actions'
+      method: POST
+      json:
+        data:
+          actionType: pause
+    response:
+      status_code: 201
+
+  - name: Wait for the delay to complete
```
Comment (SyntaxColoring) [inline robot-server/tests/integration/http_api/runs/test_pause_then_cancel.tavern.yaml]: This is referring to the "Pause the run" action above, right?

I might reword this as:

```suggestion
  - name: Wait for the first command - waitForDuration - to complete
```

* `delay` means something very different than `pause` in the Python Protocol API. This isn't the Python Protocol API, but nonetheless, being as tight as possible with terminology might help reduce confusion.
* "Wait for the [delay/pause] to complete" can be misread as "wait until the protocol is no longer [delayed/paused]," but I think what we really mean here is that we want to wait for the pause *action* to be *processed,* so the pause is, in a certain sense, active.

**Review by SyntaxColoring — COMMENTED** [review]
Nice! Here are some comments and questions on the Python side. 🐍

File: robot-server/tests/integration/http_api/runs/test_pause_then_cancel.tavern.yaml (hunk 10 above)
Comment (mcous) [reply robot-server/tests/integration/http_api/runs/test_pause_then_cancel.tavern.yaml]: > This is referring to the "Pause the run" action above, right?

@SyntaxColoring it is not, it is waiting for the first command of the protocol (a `waitForDuration`, or a "delay") to complete so that the run ends up paused with its commands in the following state:

1. `waitForDuration` - status: `succeeded`
2. `home` - status: `queued`

I've edited your suggestion accordingly to remove the ambiguity

File: robot-server/tests/integration/http_api/runs/test_pause_then_cancel.tavern.yaml (hunk 9 above)
Comment (mcous) [reply robot-server/tests/integration/http_api/runs/test_pause_then_cancel.tavern.yaml:67]: These are the statuses of the protocol's two commands

File: api/src/opentrons/protocol_engine/state/commands.py (hunk 8 above)
Comment (mcous) [reply api/src/opentrons/protocol_engine/state/commands.py:312]: The selector used to answer the question "is this JSONv6 protocol complete" is `state_view.commands.get_all_complete`. That selector checks logic checks that `state.queued_command_ids` is empty to decide if the run has finished.

Tamar identified a bug where if the stop came in at exactly the wrong time (in between two commands), the run would be left in a state where it would never transition out of `stop-requested`. This was caused by the fact that that `queued_command_ids` would never become empty if previous commands in the queue did not fail.

We didn't notice this before because usually, the `stop` request lands in the middle of a command, which will cause the command to fail due to the `asyncio.CancelledError`, clearing the rest of the queue given existing store logic

Comment (TamarZanzouri) [reply api/src/opentrons/protocol_engine/state/commands.py:312]: Should I add a comment to make this more clear? I added one but removed it.

File: robot-server/tests/integration/http_api/runs/test_pause_then_cancel.tavern.yaml (hunk 9 above)
Comment (SyntaxColoring) [reply robot-server/tests/integration/http_api/runs/test_pause_then_cancel.tavern.yaml:67]: I see. I think I'm being thrown off by the YAML syntax. No change needed, but if anyone else finds it confusing, I think a comment like this could be helpful.

```suggestion
          # Assert that there's a succeeded command followed by a queued command.
          - status: succeeded
          - status: queued
```

File: api/src/opentrons/protocol_engine/state/commands.py (hunk 8 above)
Comment (SyntaxColoring) [reply api/src/opentrons/protocol_engine/state/commands.py:312]: After that stop comes in, what are the actual and desired `status`es of each command?

This?

1. `succeeded` command
2. `succeeded` command
3. [stop action came in here]
4. `queued` command
5. `queued` command

Or this?

1. `succeeded` command
2. `succeeded` command
3. [stop action came in here]
4. `failed` command
5. `failed` command

`queued_command_ids` is just a cache/index for efficient querying. It's not the actual source of truth for which commands are queued and which ones aren't. So it seems iffy to me that it's being cleared here just by itself; I would have expected it to be paired with something that iterates through `commands_by_id` and changes everything from `queued` to `failed`, for example.

Comment (TamarZanzouri) [reply api/src/opentrons/protocol_engine/state/commands.py:312]: The status of the queued commands does not change. The run status changes. To your point we might need to clear all_command_ids as well? @mcous

Comment (mcous) [reply api/src/opentrons/protocol_engine/state/commands.py:312]: > 1. succeeded command
> 2. succeeded command
> 3. [stop action came in here]
> 4. queued command
> 5. queued command

This is the expected state of the commands after this PR / bugfix. The primary goal here is to make sure **the run itself** doesn't get locked up and is able to gracefully stop and "clean up" its hardware control resources.

> `queued_command_ids` is just a cache/index... It's not the actual source of truth for which commands are queued and which ones aren't.

This isn't quite correct. `queued_command_ids` is where the `QueueWorker` if looking to pick up execution jobs, so if `queued_command_ids` is empty, the `QueueWorker` won't execute anything. This is why `FailedCommandAction` also clears `queued_command_ids`.

> I would have expected it to be paired with something that iterates through commands_by_id and changes everything from queued to failed, for example.

I agree with this, but not for this PR. This bug fix is already a little bit of bloat in terms of the requirements of the ticket. I think our best path forward will be to introduce a `cancelled` command status. Can write that ticket up shortly.

@TamarZanzouri I don't think any more changes to command statuses are required for this PR

File: step-generation/src/utils/index.ts (hunk 7 above)
Comment (shlokamin) [reply step-generation/src/utils/index.ts]: gocha, PD uses `uuid/v1`, so i did the same when i added it to step-generation. 

i'll selfishly bump to v4 in both places so I can feel like I'm a good person who goes out of my way to make the codebase better. you're welcome, world.

File: step-generation/typings/global.d.ts [hunk 11]
```diff
@@ -0,0 +1,5 @@
+declare module 'uuid/v4' {
+  const uuidv1: any
+  // eslint-disable-next-line import/no-default-export
+  export default uuidv1
```
Comment (mcous) [inline step-generation/typings/global.d.ts]:
```suggestion
  const uuidv4: any
  // eslint-disable-next-line import/no-default-export
  export default uuidv4
```

File: step-generation/src/utils/index.ts [hunk 12]
```diff
@@ -1,17 +1,21 @@
+import uuidv1 from 'uuid/v4'
```
Comment (mcous) [inline step-generation/src/utils/index.ts]:
```suggestion
import uuidv4 from 'uuid/v4'
```

File: step-generation/src/utils/index.ts [hunk 13]
```diff
@@ -1,17 +1,21 @@
...
   orderWells,
   reduceCommandCreators,
   modulePipetteCollision,
   thermocyclerPipetteCollision,
+  isValidSlot,
 }
 export * from './commandCreatorArgsGetters'
 export * from './misc'
 export * from './heaterShakerCollision'
+export const uuid: () => string = uuidv1
```
Comment (mcous) [inline step-generation/src/utils/index.ts]:
```suggestion
export const uuid: () => string = uuidv4
```

**Review by mcous — APPROVED** [review]
🦑
